### ISSUE TITLE
BUG: `zeros_like` does not always override `dtype`

### ISSUE THREAD

**Issue by oscargus** [numpy/numpy#21603]
### Describe the issue:

According to the documentation of `zeros_like` (and related functions), providing `dtype` will override the provided data type: https://numpy.org/doc/stable/reference/generated/numpy.zeros_like.html

Basically, my interpretation is that `zeros(a.shape, dtype=foo) == zeros_like(a, dtype=foo)`.

This does not seem to be the case for `Quantity` from `astropy` (and maybe other array libraries sharing some specific mechanism).

(Which caused some issue in Astropy, see https://github.com/astropy/astropy/issues/13276, from a change in Matplotlib, see https://github.com/matplotlib/matplotlib/pull/22929. But more importantly, `zeros_like` does not appear to work as stated in the documentation. Not sure if this is a numpy or astropy issue though, but it is the numpy documentation that seems wrong.)

### Reproduce the code example:

```python
import numpy as np
import astropy.units as u

a = [1, 2, 3]*u.m

z0 = np.zeros_like(a, dtype=bool)
z1 = np.zeros(a.shape, dtype=bool)
print(repr(z0))
print(repr(z1))
```

### NumPy/Python version information:

1.22.3 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:17:03) [MSC v.1929 64 bit (AMD64)]

**Comment by oscargus** [comment]
After checking the code a bit, I realize that maybe the `like` in `zeros_like` is similar to the `like` keyword in `zeros`? In that case, I guess that the documentation should be clarified to reflect that.

**Comment by pllim** [comment]
Re: Quantity -- FWIW the full implementation is at https://github.com/astropy/astropy/blob/f0e2129aad460ea0db00730f10f98fe799aefb2f/astropy/units/quantity.py#L239

cc @mhvk

**Comment by seberg** [comment]
> After checking the code a bit, I realize that maybe the like in zeros_like is similar to the like keyword in zeros?

Yes it is.  I expect this is fully in astropy's domain.  The only thing is that `like=` is slightly restrictive, because `like=[1, 2, 3]` will not work (must already be an "array" and not something that would be converted to a NumPy array, like a list;  Although that *could* be modified).
